lib/cache/collection.go

```diff
@@ -0,0 +1,170 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package cache
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+// collection is responsible for managing a cached resource.
+type collection[T any] struct {
+	// fetcher is called by fetch to retrieve and seed the
+	// store with all known resources from upstream.
+	fetcher func(ctx context.Context, loadSecrets bool) ([]T, error)
+	// store persists all resources in memory.
+	store *store[T]
+	// watch contains the kind of resource being monitored.
+	watch types.WatchKind
+	// headerTransform is used when handling delete events in [onDelete]. Since
+	// [types.OpDelete] events only contain information about the resource key,
+	// most event handlers only emit a [types.ResourceHeader] which has enough
+	// information to identify a resource. Some resources do emit a half
+	// populated [T], or have enough information from the key to emit a full [T].
+	//
+	// If this optional transformation is supplied it will be called when
+	// processing delete events before attempting to delete the resource
+	// from the store.
```

Comment:
We should document that the indices in `store` should all work with the limited information of either the delete event coming as a `T` or with the return value of `headerTransform`, and the same is also true for the `filter` (which should maybe take a flag for deletes, or should be split into a `filterDelete`).

Reply:
Now that I think about it, isn't that a pretty harsh limitation? We are effectively limited to indices that only ever rely on the backend key, because of deletes.

Reply:
In the backend, sure? We have all the flexibility in the world here, however.

Reply:
Yeah, that's how things currently work so it doesn't really limit anything - not even the unified resource cache relies on more than what's available in the key, after all.